### PR TITLE
Early termination of result set processing

### PR DESCRIPTION
**Pull request by nezihyigitbasi**
With a custom SQL query that I have written in Tableau I noticed that the foreach loop in driver.d SQLExecuteImpl function, which traverses the result set of the query, terminates early and prevents the actual result to be returned to the application when the result data is received in the last batch retrieved from the server. The foreach construct calls the statement client's empty(), front(), and popfront() functions in order as expected, but the empty() function returns true before the last result (that contains the actual result of the query) is retrieved & processed with the front() call.

### CODE REVIEW

**Comment by nezihyigitbasi** [comment]
Sometimes the data array in the response JSON is available only in the last batch (I guess for small tables, but not sure about it) and current statement client doesn't handle that case. Here are some logs I added, only the last batch contains the data array (lines starting with popFront() print the response JSON ) and empty() evaluates to true right after that and prevents the processing of the result data:

```
2014-Aug-21 09:27:48.3482472           driver\driver.d:435   SQLExecuteImpl SELECT *
FROM (
  select id, cast(flag as varchar) as flag_ from test_bool
) "TableauSQL"
LIMIT 1 
2014-Aug-21 09:27:49.2102444  client\statementclient.d:150   empty()  false
2014-Aug-21 09:27:49.2112481  client\statementclient.d:137   front()
2014-Aug-21 09:27:49.2112481           driver\driver.d:450   SQLExecute working on result batch 1
2014-Aug-21 09:27:49.2112481  client\statementclient.d:150   empty()  false
2014-Aug-21 09:27:50.8052446  client\statementclient.d:144   popFront()  {"id":"20140821_092927_00217_3dasc","infoUri":"http://10.64.24.11:8080/v1/query/20140821_092927_00217_3dasc","partialCancelUri":"http://10.64.24.11:8080/v1/stage/20140821_092927_00217_3dasc.1","nextUri":"http://10.64.24.11:8080/v1/statement/20140821_092927_00217_3dasc/2","columns":[{"name":"id","type":"bigint"},{"name":"flag_","type":"varchar"}],"stats":{"state":"RUNNING","scheduled":false,"nodes":1,"totalSplits":1,"queuedSplits":0,"runningSplits":1,"completedSplits":0,"userTimeMillis":0,"cpuTimeMillis":0,"wallTimeMillis":0,"processedRows":0,"processedBytes":0,"rootStage":{"stageId":"0","state":"RUNNING","done":false,"nodes":1,"totalSplits":1,"queuedSplits":0,"runningSplits":1,"completedSplits":0,"userTimeMillis":0,"cpuTimeMillis":0,"wallTimeMillis":0,"processedRows":0,"processedBytes":0,"subStages":[{"stageId":"1","state":"SCHEDULING","done":false,"nodes":0,"totalSplits":0,"queuedSplits":0,"runningSplits":0,"completedSplits":0,"userTimeMillis":0,"cpuTimeMillis":0,"wallTimeMillis":0,"processedRows":0,"processedBytes":0,"subStages":[]}]}}}

2014-Aug-21 09:27:50.8062513  client\statementclient.d:150   empty()  false
2014-Aug-21 09:27:50.8062513  client\statementclient.d:137   front()
2014-Aug-21 09:27:50.8062513           driver\driver.d:450   SQLExecute working on result batch 2
2014-Aug-21 09:27:50.8062513  client\statementclient.d:150   empty()  false
2014-Aug-21 09:27:51.3902322  client\statementclient.d:144   popFront()  {"id":"20140821_092927_00217_3dasc","infoUri":"http://10.64.24.11:8080/v1/query/20140821_092927_00217_3dasc","columns":[{"name":"id","type":"bigint"},{"name":"flag_","type":"varchar"}],"data":[[1,"true"]],"stats":{"state":"FINISHED","scheduled":true,"nodes":1,"totalSplits":2,"queuedSplits":0,"runningSplits":0,"completedSplits":2,"userTimeMillis":2,"cpuTimeMillis":2,"wallTimeMillis":2,"processedRows":2,"processedBytes":15,"rootStage":{"stageId":"0","state":"FINISHED","done":true,"nodes":1,"totalSplits":1,"queuedSplits":0,"runningSplits":0,"completedSplits":1,"userTimeMillis":1,"cpuTimeMillis":1,"wallTimeMillis":1,"processedRows":1,"processedBytes":11,"subStages":[{"stageId":"1","state":"FINISHED","done":true,"nodes":1,"totalSplits":1,"queuedSplits":0,"runningSplits":0,"completedSplits":1,"userTimeMillis":0,"cpuTimeMillis":0,"wallTimeMillis":0,"processedRows":2,"processedBytes":15,"subStages":[]}]}}}

2014-Aug-21 09:27:51.3902322  client\statementclient.d:150   empty()  true
2014-Aug-21 09:27:51.3902322  client\statementclient.d:158   Pre del
```

**Comment by nezihyigitbasi** [comment]
The query works fine from the presto-cli.

```
presto:nyigitbasi> select id, cast(flag as varchar) as flag_ from test_bool;
 id | flag_
----+-------
  1 | true
  2 | false
(2 rows)

Query 20140821_093734_00223_3dasc, FINISHED, 1 node
Splits: 2 total, 2 done (100.00%)
0:02 [2 rows, 15B] [0 rows/s, 7B/s]

presto:nyigitbasi>
```

File: client/statementclient.d [hunk 1]
```diff
@@ -131,18 +131,23 @@ struct StatementClient {
...
+    immutable(QueryResults) front() nothrow {
+        if (results_.nextURI == "" && results_.succeeded) {
+            processedAll_ = true;
+        }
         return results_;
     }
 
     void popFront() {
-        assert(!empty);
+        if(empty){
```
Comment (andralex) [inline client/statementclient.d]: I think the prevalent style in this project is `if (empty) {` i.e. spaces after the `if` and before the `{`. Also, if client code tries to `popFront` from an empty range it's in error. Did you mean this?

``` d
if (processedAll_) {
    return;
}
assert(!empty);
```

That takes care of the corner case and still makes sure caller code is correct.

Comment (nezihyigitbasi) [inline client/statementclient.d]: Thanks Andrei. I updated the code with your proposal as it is functionally equivalent with an additional assert statement. The reason I didn't use an assertion initially was http get call also has an assertion that was failing for an empty url.

**Comment by andralex** [comment]
I need to push back on this. A range should not call front to determine whether it's empty. The range protocol goes as you mention like this:
1. create range
2. check for empty
3. If empty returns true then nothing else works
4. Otherwise (empty returns false), can call front and popFront.
5. Repeat from 2

The flag `processedAll_` looks like a good approach, but it shouldn't be set inside front; instead, the constructor, popFront, or empty should set it.

Could you try a small refactoring? I'd be glad to try it but I'm not currently rigged with a Windows machine and data connectivity.

**Comment by nezihyigitbasi** [comment]
@andralex Refactored to set `processedAll_` in `popFront()`  & tested.

**Comment by andralex** [comment]
Great, thanks!

**Comment by andralex** [comment]
merged

**Pull request closed without merging**